src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultImportDocumentManager.cs

```diff
@@ -0,0 +1,235 @@
+
+            return imports;
+        }
+
+        private void FireImportChanged(string importPath, ImportChangeKind kind)
+        {
+            var handler = Changed;
+            if (handler != null && _importTrackerCache.TryGetValue(importPath, out var importTracker))
+            {
+                var args = new ImportsChangedEventArgs(importPath, kind, importTracker.AssociatedDocuments);
+                handler(this, args);
+            }
+        }
+
+        private void OnFilesChanged(uint fileCount, string[] filePaths, uint[] fileChangeFlags)
+        {
+            for (var i = 0; i < fileCount; i++)
+            {
+                var kind = ImportChangeKind.Changed;
+                var flag = (_VSFILECHANGEFLAGS)fileChangeFlags[i];
+
+                if (flag.HasFlag(_VSFILECHANGEFLAGS.VSFILECHG_Del))
+                {
+                    kind = ImportChangeKind.Removed;
+                }
+                else if (flag.HasFlag(_VSFILECHANGEFLAGS.VSFILECHG_Add))
+                {
+                    kind = ImportChangeKind.Added;
+                }
+
+                FireImportChanged(filePaths[i], kind);
+            }
+        }
+
+        private void StartListeningForChanges(ImportTracker importTracker)
+        {
+            try
+            {
+                if (importTracker.FileChangeCookie == VSConstants.VSCOOKIE_NIL)
+                {
+                    Marshal.ThrowExceptionForHR(
+                        _fileChangeService.AdviseFileChange(
+                            importTracker.FilePath,
+                            FileChangeFlags,
+                            new ImportDocumentEventSink(this, _foregroundDispatcher),
+                            out var cookie));
+
+                    importTracker.FileChangeCookie = cookie;
+                }
+            }
+            catch (Exception exception)
+            {
+                _errorReporter.ReportError(exception);
+            }
+        }
+
+        private void StopListeningForChanges(ImportTracker importTracker)
+        {
+            try
+            {
+                if (importTracker.FileChangeCookie != VSConstants.VSCOOKIE_NIL)
+                {
+                    Marshal.ThrowExceptionForHR(_fileChangeService.UnadviseFileChange(importTracker.FileChangeCookie));
+                    importTracker.FileChangeCookie = VSConstants.VSCOOKIE_NIL;
+                }
+            }
+            catch (Exception exception)
+            {
+                _errorReporter.ReportError(exception);
+            }
+        }
+
+        private class ImportTracker
+        {
+            public ImportTracker(string filePath)
+            {
+                FilePath = filePath;
+                FileChangeCookie = VSConstants.VSCOOKIE_NIL;
+            }
+
+            public string FilePath { get; }
+
+            public HashSet<string> AssociatedDocuments { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```

Comment:
nit: Try and group all initializations. Aka, move this init into the ctor since your initializing everything else there.